### PR TITLE
Add tests of WCS accuracy.

### PR DESCRIPTION
**Pull request by schlafly**
This PR adds a new regression test demonstrating that we can fit the WCS of images to <1.3 / sqrt(2) mas.  This is motivated by science requirements demanding that different images must agree to within 1.3 mas.  We show instead that our individual images agree with absolute truth to 1.3 / sqrt(2), since if each image agrees at 1.3 / sqrt(2) mas, then they will agree with one another at 1.3 mas.

There are a few different types of changes in here:
1. The new tests.  I use the existing test_tweakreg regression test, but update the file it uses to a new "shifted" file.  This is identical to the standard WFI01 imaging file we use for everything, but where I have inserted an error into ra_ref and dec_ref in the wcsinfo and updated the assigned WCS accordingly, following the same procedure that we do for the "repointed" regression test file.  The code for the new regression test file has been added to make_regtestdata.sh.  I remove the old single-file association file for the tweakreg regression test as it does not seem to serve a purpose.  <br />
The core of the new tests is comparing the simulated WCS with the tweakreg'd WCS.  We grab the true WCS from the 'romanisim' extension of the uncalibrated asdf file, and compare it with the WCS of the tweakreg output.  We make a grid of points on the detector, convert them to sky coordinates with both sets of WCS, and compute the RMS difference between the sky coordinates in milliarcseconds.  The test passes if the resulting RMS is less than 1.3 / sqrt(2) mas.
2. The update to make_regtestdata.sh needed to make the new regression test files.
3. The source detection step is updated to use fit_psf by default, so that we get the astrometric accuracy on the individual detections we need in order to match to Gaia.
4. We change the name of the output extension when using PSF source detection to match the name without it; both of these options now produce a source_detection.tweakreg_catalog output.  We should consider different options there when eventually making L2 files, but this generates the fewest changes now.
5. It turns on "clip_accum" in the when calling tweakwcs.align_wcs, since this is more robust than the default clip_accum = False.
6. It adds a workaround for https://github.com/astropy/astropy/issues/16055 , which otherwise leads to a "unit mismatch" in source detection, since apparently DN / s and DN ^ 1.0 / s ^ 1.0 aren't quite the same quantity.

This requires a new set of L1 files incorporating the changes from
https://github.com/spacetelescope/romanisim/pull/110
and won't pass unit tests without them.  Since that requires updating the "main" uncal file on which many regtests rely, I haven't done that yet.  Here is the log from a local test run:

```
TEST_BIGDATA=~/projects/roman/data/regtest/ pytest romancal/regtest --bigdata --basetemp blah -k test_tweakreg -s
========================================================== test session starts ===========================================================
platform darwin -- Python 3.12.2, pytest-7.4.4, pluggy-1.4.0
rootdir: /Users/eschlafly/projects/roman/romancal
configfile: pyproject.toml
plugins: ci_watson-0.6.2, astropy-0.11.0, remotedata-0.4.1, jwst-1.14.0, env-1.1.3, cov-4.1.0, hypothesis-6.99.9, filter-subpackage-0.2.0, doctestplus-1.2.1, astropy-header-0.2.2, mock-3.12.0, arraydiff-0.6.1, xdist-3.5.0, openfiles-0.5.0, romancal-0.14.1.dev101+g3341389.d20240320, asdf-3.1.0
collected 42 items / 41 deselected / 1 selected                                                                                          

romancal/regtest/test_tweakreg.py 2024-04-17 14:35:28,441 - stpipe.TweakRegStep - INFO - TweakRegStep instance created.
2024-04-17 14:35:29,344 - CRDS - ERROR -  Error determining best reference for 'pars-tweakregstep'  =   Unknown reference type 'pars-tweakregstep'
2024-04-17 14:35:29,345 - stpipe.TweakRegStep - INFO - TweakRegStep instance created.
2024-04-17 14:35:29,346 - stpipe - INFO - Hostname: darkshaa2022.stsci.edu
2024-04-17 14:35:29,346 - stpipe - INFO - OS: Darwin
2024-04-17 14:35:29,437 - stpipe.TweakRegStep - INFO - Step TweakRegStep running with args ('/Users/eschlafly/projects/roman/romancal/blah/test_tweakreg0/r0000101001001001001_01101_0001_WFI01_shift_cal.asdf',).
2024-04-17 14:35:29,440 - stpipe.TweakRegStep - INFO - Step TweakRegStep parameters are:
  pre_hooks: []
  post_hooks: []
  output_file: /Users/eschlafly/projects/roman/romancal/blah/test_tweakreg0/r0000101001001001001_01101_0001_WFI01_shift_tweakregstep.asdf
  output_dir: None
  output_ext: .asdf
  output_use_model: True
  output_use_index: True
  save_results: True
  skip: False
  suffix: tweakregstep
  search_output_file: True
  input_dir: /Users/eschlafly/projects/roman/romancal/blah/test_tweakreg0
  use_custom_catalogs: False
  catalog_format: ascii.ecsv
  catfile: ''
  catalog_path: ''
  enforce_user_order: False
  expand_refcat: False
  minobj: 15
  searchrad: 2.0
  use2dhist: True
  separation: 1.0
  tolerance: 0.7
  fitgeometry: rshift
  nclip: 3
  sigma: 3.0
  abs_refcat: GAIADR3
  save_abs_catalog: False
  abs_minobj: 15
  abs_searchrad: 6.0
  abs_use2dhist: True
  abs_separation: 0.1
  abs_tolerance: 0.7
  abs_fitgeometry: rshift
  abs_nclip: 3
  abs_sigma: 3.0
2024-04-17 14:35:29,440 - stpipe.TweakRegStep - INFO - All source catalogs will be saved to: /Users/eschlafly/projects/roman/romancal/blah/test_tweakreg0
2024-04-17 14:35:29,686 - stpipe.TweakRegStep - INFO - Detected 114 sources in r0000101001001001001_01101_0001_WFI01_shift_cal.asdf.
2024-04-17 14:35:29,686 - stpipe.TweakRegStep - INFO - 
2024-04-17 14:35:29,686 - stpipe.TweakRegStep - INFO - Number of image groups to be aligned: 1.
2024-04-17 14:35:29,686 - stpipe.TweakRegStep - INFO - Image groups:
2024-04-17 14:35:29,707 - stpipe.TweakRegStep - INFO - * Images in GROUP 'r0000101001001001001_01101_0001_WFI01_shift_cal':
2024-04-17 14:35:29,707 - stpipe.TweakRegStep - INFO -      r0000101001001001001_01101_0001_WFI01_shift_cal
2024-04-17 14:35:29,707 - stpipe.TweakRegStep - INFO - 
2024-04-17 14:35:30,563 - stpipe.TweakRegStep - INFO -  
2024-04-17 14:35:30,563 - stpipe.TweakRegStep - INFO - ***** tweakwcs.imalign.align_wcs() started on 2024-04-17 14:35:30.563057
2024-04-17 14:35:30,563 - stpipe.TweakRegStep - INFO -       Version 0.8.8.dev5+gfb46242
2024-04-17 14:35:30,563 - stpipe.TweakRegStep - INFO -  
2024-04-17 14:35:30,622 - stpipe.TweakRegStep - INFO - Aligning image catalog 'GROUP ID: 987654' to the reference catalog.
2024-04-17 14:35:30,629 - stpipe.TweakRegStep - INFO - Matching sources from 'r0000101001001001001_01101_0001_WFI01_shift_cal' catalog with sources from the reference 'Unnamed' catalog.
2024-04-17 14:35:30,630 - stpipe.TweakRegStep - INFO - Computing initial guess for X and Y shifts...
2024-04-17 14:35:30,631 - stpipe.TweakRegStep - INFO - Found initial X and Y shifts of 0.54, 1.086 (arcsec) with significance of 105 and 106 matches.
2024-04-17 14:35:30,631 - stpipe.TweakRegStep - INFO - Found 107 matches for 'GROUP ID: 987654'...
2024-04-17 14:35:30,632 - stpipe.TweakRegStep - INFO - Performing 'rshift' fit
2024-04-17 14:35:30,634 - stpipe.TweakRegStep - INFO - Computed 'rshift' fit for GROUP ID: 987654:
2024-04-17 14:35:30,634 - stpipe.TweakRegStep - INFO - XSH: -0.406646  YSH: -1.00039    ROT: 0.000259275    SCALE: 1
2024-04-17 14:35:30,634 - stpipe.TweakRegStep - INFO - 
2024-04-17 14:35:30,634 - stpipe.TweakRegStep - INFO - FIT RMSE: 0.00317382   FIT MAE: 0.00221558
2024-04-17 14:35:30,634 - stpipe.TweakRegStep - INFO - Final solution based on 103 objects.
2024-04-17 14:35:30,668 - stpipe.TweakRegStep - INFO -  
2024-04-17 14:35:30,668 - stpipe.TweakRegStep - INFO - ***** tweakwcs.imalign.align_wcs() ended on 2024-04-17 14:35:30.668135
2024-04-17 14:35:30,668 - stpipe.TweakRegStep - INFO - ***** tweakwcs.imalign.align_wcs() TOTAL RUN TIME: 0:00:00.105078
2024-04-17 14:35:30,668 - stpipe.TweakRegStep - INFO -  
2024-04-17 14:35:31,574 - stpipe.TweakRegStep - INFO - Saved model in r0000101001001001001_01101_0001_WFI01_shift_tweakregstep.asdf
2024-04-17 14:35:31,574 - stpipe.TweakRegStep - INFO - Step TweakRegStep done
2024-04-17 14:35:31,896 - stpipe.TweakRegStep - INFO - DMS280 MSG: TweakReg step recorded as complete? : True
2024-04-17 14:35:31,896 - stpipe.TweakRegStep - INFO - DMS280 MSG: TweakReg created new attribute with fit results? :            True
2024-04-17 14:35:31,896 - stpipe.TweakRegStep - INFO - DMS280 MSG: TweakReg created new coordinate frame 'v2v3corr'? :            True
2024-04-17 14:35:48,209 - stpipe.TweakRegStep - INFO - DMS280 MSG: Was the proper TweakReg data produced? : True
2024-04-17 14:35:48,413 - stpipe.TweakRegStep - INFO - DMS488 MSG: WCS agrees with true WCS to  0.38 mas, less than 1.3 / sqrt(2)?  PASS
2024-04-17 14:35:48,414 - stpipe.TweakRegStep - INFO - DMS405 MSG: WCS agrees with true WCS to  0.38 mas, less than 5 / sqrt(2)?  PASS
2024-04-17 14:35:48,433 - stpipe - INFO - METRIC - DMS280, DMS405, DMS488 - test_tweakreg - PASS
.

============================================================ warnings summary ============================================================
<clip>
============================================== 1 passed, 41 deselected, 1 warning in 22.94s ==============================================
```

The final agreement is quite good, 0.38 mas RMS for this test.  At some point we should run a more complete set of filters and locations on the sky, but not in these regression tests.

**Checklist**
- [ ] added entry in `CHANGES.rst` under the corresponding subsection
- [X] updated relevant tests
- [X] updated relevant milestone(s)
- [X] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below. [How to run regression tests on a PR](https://github.com/spacetelescope/romancal/wiki/Running-Regression-Tests-Against-PR-Branches)

### CODE REVIEW

File: romancal/regtest/test_tweakreg.py [hunk 1]
```diff
@@ -64,3 +66,24 @@ def test_tweakreg(rtdata, ignore_asdf_paths, tmp_path):
         f"DMS280 MSG: Was the proper TweakReg data produced? : {diff.identical}"
     )
     assert diff.identical, diff.report()
+
+    wcstweak = tweakreg_out.meta.wcs
+    orig_model_asdf = asdf.open(orig_uncal)
+    wcstrue = orig_model_asdf["romanisim"]["wcs"]  # simulated, true WCS
```
Comment (nden) [inline romancal/regtest/test_tweakreg.py:72]: This is the uncal cube with attached 2D WCS, generated manually?
There are different ways to generate a "true" WCS, is the process described somewhere?

File: romancal/flatfield/flat_field.py [hunk 2]
```diff
@@ -121,6 +121,8 @@ def apply_flat_field(science, flat):
         science.var_flat = science.data**2 / flat_data_squared * flat_err**2
 
     science.err = np.sqrt(science.var_poisson + science.var_rnoise + science.var_flat)
+    science.err = science.err.to(science.data.unit)
+    # Workaround for https://github.com/astropy/astropy/issues/16055
```
[the remark this answers is not in the document]
Comment (schlafly) [reply romancal/flatfield/flat_field.py:125]: Thanks Nadia.  Let me know which of those options you prefer.  astropy 6.0.1 was a little more cutting edge than I wanted to require to pull in this one liner, but I agree it's an option.

File: romancal/regtest/test_tweakreg.py [hunk 3]
```diff
@@ -8,25 +11,24 @@
 from .regtestdata import compare_asdf
 
 
-@metrics_logger("DMS280")
+@metrics_logger("DMS280", "DMS405", "DMS488")
 @pytest.mark.bigdata
 def test_tweakreg(rtdata, ignore_asdf_paths, tmp_path):
     # N.B.: the data were created using WFIsim and processed through
```
[the remark this answers is not in the document]
Comment (schlafly) [reply romancal/regtest/test_tweakreg.py]: Thanks.  No, the comment was out of date.  I've updated it.

File: romancal/regtest/test_tweakreg.py (hunk 1 above)
Comment (schlafly) [reply romancal/regtest/test_tweakreg.py:71]: Every current file that romanisim makes gets a special 'romanisim' extension at the root of the asdf tree.  This extension contains the WCS that was used to simulate the image.  I'm comparing that true WCS (true, in the sense that it was the WCS actually used to simulate the image) to the WCS that tweakreg recovered.

The process used to generate a simulator WCS is the same as the process used for AssignWCS in romancal.  It's just CRDS distortion maps + shifts and scales + rotations on the sky.  The only systematic difference between the simulator WCS and the default AssignWCS WCS in romancal for these files comes from the fact that we have inserted a small error into the wcsinfo in the "shift" files; tweakreg is responsible for finding and correcting this error.

Does that sound right to you?

**Review by nden — APPROVED** [review]

**Pull request merged**